ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -340,3 +372,140 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
     builder = patSynBuilder p
     field_labels = patSynFieldLabels p
     orig_args' = map scaledThing orig_args
+
+-- --------------------------------------------------------------------------------
+
+-- | A binding expression with its id and location.
+data WhereBinding = WhereBinding
+    { bindingId  :: Id
+    -- ^ Each WhereBinding represents a id in binding expression.
+    , bindingLoc :: SrcSpan
+    -- ^ Location for a individual binding in a pattern.
+    -- Here we use the this and offset to render the type signature at the proper place.
+    , offset     :: Int
+    -- ^ Column offset between whole binding and individual binding in a pattern.
+    --
+    -- Example: For @(a, b) = (1, True)@, there will be two `WhereBinding`s:
+    -- - `a`: WhereBinding id_a loc_a 0
+    -- - `b`: WhereBinding id_b loc_b 4
+    }
+
+-- | Existed bindings in a where clause.
+data WhereBindings = WhereBindings
+  { bindings        :: [WhereBinding]
+  , existedSigNames :: [Name]
+  -- ^ Names of existing signatures.
+  -- It is used to hide type lens for existing signatures.
+  --
+  -- NOTE: The location of this name is equal to
+  -- the binding name.
+  --
+  -- Example:
+  -- @
+  -- f :: Int
+  -- f = 42
+  -- @
+  -- The location of signature name `f`(first line) is equal to
+  -- the definition of `f`(second line).
+  }
+
+-- | All where clauses from type checked source.
+findWhereQ :: GenericQ [HsLocalBinds GhcTc]
+findWhereQ = everything (<>) $ mkQ [] (pure . findWhere)
+  where
+    findWhere :: GRHSs GhcTc (LHsExpr GhcTc) -> HsLocalBinds GhcTc
+    findWhere = grhssLocalBinds
```

Comment:
This looks very expensive, don't we traverse the full AST each time the client requests InlayHints in this case? I think this needs to be tucked away in a rule, so that we at least don't recompute when nothing changes.

Do you happen to know when the client is supposed to ask for inlay hints?

Reply:
Yes, this should be tucked away in the rule.

I think editor will request InlayHints every time the buffer text changes.